File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> mp;
        int left = 0, right = 0;
        int maxLen = 0;
        for(right = 0;right < s.length();right++){
            mp[s[right]]++;

            while(mp[s[right]]>1){
                mp[s[left]]--;
                left++; // shrink
            }
            maxLen = max(maxLen, right - left + 1);
        }return maxLen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> seen = new HashSet<>();
        int l = 0;
        int ans = 0;
        for (int r = 0; r < s.length(); r++) {
            while (seen.contains(s.charAt(r))) {
                seen.remove(s.charAt(l));
                l++;
            }
            seen.add(s.charAt(r));
            ans = Math.max(ans, r - l + 1);
        }
        return ans;
        /*
        HashMap<Character, Integer> map = new HashMap<>();
        int l = 0;
        int ans = 0;

        for (int r = 0; r < s.length(); r++) {
            char curr = s.charAt(r);

            // Agar duplicate mila, toh 'l' ko direct aage jump karwao
            if (map.containsKey(curr)) {
                l = Math.max(l, map.get(curr) + 1);
            }

            // Map mein current character ka index update karo
            map.put(curr, r);

            // Answer update karo
            ans = Math.max(ans, r - l + 1);
        }

        return ans;
        */
    }
}
